keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt

```diff
@@ -382,11 +389,106 @@ class ClusterHandler(
     return spec
   }
 
-  private fun List<ResourceDiff<ServerGroup>>.createsNewServerGroups() =
-    any {
-      !it.isCapacityOrAutoScalingOnly()
+  override suspend fun actuationInProgress(resource: Resource<ClusterSpec>) =
+    resource
+      .spec
+      .locations
+      .regions
+      .map { it.name }
+      .any { region ->
+        orcaService
+          .getCorrelatedExecutions("${resource.id}:$region")
+          .isNotEmpty()
+      }
+
+  /**
+   * Retrieve server group entity tags from clouddriver to identity the "source" of the server group (pipeline or
+   * orchestration), then retrieve execution details from orca to determine the deployment strategy used for that
+   * server group.
+   */
+  private fun ServerGroup.discoverDeploymentStrategy(): ClusterDeployStrategy? {
+    val entityTags = runBlocking {
+      log.debug("Looking for entity tags on server group $name in application ${moniker.app}, " +
+        "account ${location.account} in search of pipeline/task correlation.")
+      cloudDriverService.getEntityTags(
+        cloudProvider = "aws",
+        account = location.account,
+        application = moniker.app,
+        entityType = "servergroup",
+        entityId = name
+      )
+    }
+
+    if (entityTags.isEmpty()) {
+      log.warn("Unable to find entity tags for server group $name in application ${moniker.app}, " +
+        "account ${location.account}.")
+      return null
     }
 
+    val spinnakerMetadata = entityTags.first().tags
+      .find { it.name == "spinnaker:metadata" }
+      ?.let {
+        if (it.value !is Map<*, *>) {
+          null
+        } else {
+          it.value as Map<*, *>
+        }
+      }
+
+    if (spinnakerMetadata == null ||
+      spinnakerMetadata["executionType"] == null ||
+      spinnakerMetadata["executionId"] == null
+    ) {
+      log.warn("Unable to find Spinnaker metadata for server group $name in application ${moniker.app}, " +
+        "account ${location.account} in entity tags.")
+      return null
+    }
+
+    val executionType = spinnakerMetadata["executionType"]!!.toString()
+    val executionId = spinnakerMetadata["executionId"]!!.toString()
+    val execution = runBlocking {
+      when (executionType) {
+        "orchestration" -> orcaService.getOrchestrationExecution(executionId)
+        "pipeline" -> orcaService.getPipelineExecution(executionId)
+        else -> throw SystemException("Unsupported execution type $executionType in Spinnaker metadata.")
+      }
+    }
+
+    // get context from the appropriate execution stage for the execution type, drilling down into the data as needed
+    val context = if (executionType == "pipeline") {
```

Comment:
This set of statements finds the right spot in the data structure (which is different between a task execution and a pipeline execution) to extract the context data we need.